stdlib/lang-parser/parserModule/bufferReader.bal

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2019 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/io;
+import ballerina/log;
+
+public type BufferReader object {
+    //size of the character array
+    private int capacity;
+    //pointer which reads the character array
+    private int readPointer = 0;
+    private io:ReadableCharacterChannel sourceChannel;
+    //array which stores the characters
+    private string[] characterArray = [];
+    boolean isEof = false;
+    //counter to fill the buffer
+    private int bufferSize = 0;
+
+    public function __init(int capacity = 5,io:ReadableCharacterChannel sourceChannel) {
+        self.capacity = capacity;
+        self.sourceChannel = sourceChannel;
+        var result = self.fillBuffer();
+
+        if (result is error) {
+            // Doing this right now as we cannot handle errors returned from constructor
+            panic result;
+        }
+    }
+    //buffer is filled character by character based on the capacity defined
+
+    function fillBuffer() returns error? {
+        while (self.bufferSize < self.capacity) {
+            self.characterArray[self.bufferSize] = check self.sourceChannel.read(1);
```

Comment:
Is this the best way to read the char channel? Wouldn't this be expensive to read one char at a time? @wggihan thoughts?

Reply:
@pamod What was the expected behavior in this case when we initially design this?

Reply:
Character channel allows you to read a chunk. However in this use case you're trying to populate the buffer array which might require you to to an iteration anyhow. 

However, I/O calls can be reduced. 

So i propose do the following,

1. Read all the characters self.sourceChannel.read(self.capacity-1); which will return the string 
2. If the number of characters are < self.capacity-1 it wouldn't matter, it will return what ever which is there
3. Then cast the string into char[] or iterate through each character in the string and assign it to the buffer. 

This will reduce the I/O overhead which will be created per iteration.

Reply:
When characters are read as a chunk, it doesn't read the `EOF` character.
 But the `EOF` character is needed as I return it as a token. Also the last whitespace characters of the source code is attached to the `EOF token`